### PR TITLE
Enhance `library folder contents` search API

### LINKED ISSUE #14180: On Dataset Libraries Page, Search feature frozen on zero results
**Describe the bug**
On the "Shared Data" >> "Dataset Libraries" Page, the Search input field results in "Loading..." icon spinning for 30+ seconds.

**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: 22.05 on https://usegalaxy.org/libraries

**To Reproduce**
1. Go to https://usegalaxy.org/libraries
2. Click on "2019_nCoV" (first item in list for US-version) >> "Sequencing datasets" >> "ILLUMINA"
3. In Search input field, type: sff
4. See "Loading..." for 30+ seconds

**Expected behavior**
I would expect that if there were no results returned, it would stop the "Loading..." from appearing and just saying something simple like "no results"

**Screenshots**
Greenshot 2022-06-23 13 35 14

### PR DESCRIPTION
**Pull request by davelopez**
Fixes #14180

This PR moves the database logic from the service layer to the manager (where it should live) and tries optimizing the index endpoint in the library folder contents API using a database query instead of retrieving all contents and performing manual filtering.

Includes:
- A bunch of refactorings to simplify the endpoint logic.
- Increase the test coverage.
- Make search text case insensitive.
- Search query optimization. See next section.
- Fix `dataset_is_private_to_user` raising an exception when the user is a bootstrap admin.

### TLDR
Request time reduction from **2028.7 ms** to **87.6 ms** searching on a folder with 3000 datasets (around 20x faster) :tada:

### Details
In my local setup, I ran a simple postman test 20 times (with a delay of 100ms between requests) and measured the response time on:
```
http://127.0.0.1:8080/api/folders/F33b43b4e7093c91f/contents?include_deleted=false&limit=10&offset=0&search_text=300
```
The folder `F33b43b4e7093c91f` contains 3000 datasets (2700 non-deleted) and the `search_text=300` should yield 4 datasets as result.

On `dev` branch the response times in milliseconds were:
```json
{
  "times": [2011, 2172, 2136, 1797, 1885, 2173, 1860, 1972, 1849, 2172, 1777, 1916, 2240, 2405, 2337, 2051, 1839, 2103, 1861, 2018],
  "count": 20,
  "totalTime": 40574,
}
```
Which results in an average response time of **2028.7 ms**

On `this` branch the response times in milliseconds were:
```json
{
  "times": [74, 69, 73, 113, 69, 69, 127, 116, 87, 129, 109, 67, 68, 100, 72, 65, 66, 69, 85, 125],
  "count": 20,
  "totalTime": 1752,
}
```
Which results in an average response time of **87.6 ms** :tada: 

## TODO
- [x] Fix permission check for private and public datasets in query
- [x] Additional refactorings
- <del>Support `order by`. The client right now only orders the current page.</del> I will add this in a follow-up PR to keep the scope of this PR limited. Tracked here #14256

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  - Go to `Shared Data -> Data Libraries`
  - Find a folder with a large number of datasets
  - Try to search by entering any text

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Comment by davelopez** [comment]
Getting there, but there's still something off with the new query... I'm getting more `total counts` than I should when including deleted items, probably I'm doing the joining wrong... I'm trying to learn SQLAlchemy in the process but it looks like I don't have the skill yet :laughing: so any hints and recommendations would be much appreciated.

**Comment by davelopez** [comment]
Ok, I think now is ready for review!

**Comment by davelopez** [comment]
This test https://github.com/galaxyproject/galaxy/runs/7187272372?check_suite_focus=true#step:9:808 is failing consistently in CI while it passes consistently locally (and in a fresh checkout) for me. I even tried it both with SQLite and Postgres databases just in case. 
Does anybody have an idea what may be the problem here?

**Comment by davelopez** [comment]
API tests are finally green! :tada:

**Review by dannon — APPROVED** [review]
